app/lib/actions/middleware/watch_delegated_proxy_sub_tasks.rb

```diff
@@ -0,0 +1,73 @@
+module Actions
+  module Middleware
+    class WatchDelegatedProxySubTasks < ::Dynflow::Middleware
+      class CheckOnProxyActions; end
+      POLL_INTERVAL = 30
+
+      def run(event = nil)
+        if event.nil?
+          set_clock
+        elsif event == CheckOnProxyActions
+          poll_proxy_tasks
+          set_clock
+          action.send(:suspend)
+        end
+        pass event
+      end
+
+      private
+
+      def set_clock
+        action.world.clock.ping action.send(:suspended_action),
+                                POLL_INTERVAL,
+                                CheckOnProxyActions
+      end
+
+      def poll_proxy_tasks
+        results = delegated_actions_by_proxy.map { |(url, actions)| poll_proxy_actions(url, actions) }.flatten
+
+        missing, present = results.partition { |result| result[:result].nil? }
+        if missing.any?
+          event = ::Actions::ProxyAction::ProxyActionMissing.new
+          notify(event, missing.map { |action| action[:action] })
+        end
+
+        stopped = present.select { |action| %w[stopped paused].include? action[:result]['state'] }
+        if stopped.any?
+          event = ::Actions::ProxyAction::ProxyActionStopped.new
+          notify(event, stopped.map { |a| a[:action] })
+        end
+      end
+
+      def notify(event, actions)
+        actions.each do |action|
+          action.world.event action.execution_plan_id,
+                             action.run_step_id,
+                             event
+        end
+      end
+
+      def delegated_actions
+        action.sub_plans('state' => %w[running])
```

Comment:
This can be potentially expensive and memory-hungry method, loading all the actions data for a job with 10000s of hosts. I'm thinking of an alternative approach, where we would keep the track of association between local action and remote one in a separate table, where we would track the relation between local step and remote task. We could also use this relation to group the requests to the proxy when creating the remote actions.

In short, instead of calling to proxy on every host, we would just mark that the remote action is ready to be triggered at the remote host, and a periodic procedure (that could be done as part of the parent task) would collect the info from all actions that would be ready to be triggered remotely and trigger it in bulk.

I think we could do this in two phases, leaving the bulk triggering as a separate PR, but we could do the work in the PR ready for moving this direction.